python/python/bystro/tada_mixture/mixture_tada_em.py

```diff
@@ -0,0 +1,394 @@
+        ----------
+        data : np.array-like,shape=(N_variants,n_phenotypes)
+            The data of counts, variants x phenotypes
+
+        Returns
+        -------
+        z_hat : np.array-like,shape=(N_samples,)
+            The cluster identities
+        """
+        N = data.shape[0]
+        log_proba = np.zeros((N, self.K))
+        for i in range(N):
+            for k in range(self.K):
+                log_proba[i, k] = np.sum(
+                    st.poisson.logpmf(data[i], self.Lambda[k])
+                )
+        return log_proba
+
+    def _fill_training_options(self, training_options):
+        """
+        This fills any relevant parameters for the learning algorithm
+
+        Parameters
+        ----------
+        training_options : dict
+
+        Returns
+        -------
+        tops : dict
+        """
+        default_options = {"n_iterations": 100}
+        tops = deepcopy(default_options)
+        tops.update(training_options)
+        return tops
+
+
+class MVTadaZPoissonEM(object):
+    def __init__(self, K=4, training_options={}):
+        """
+        This is a Product ZIP latent variable model.
+
+        Parameters
+        ----------
+        K : int,default=4
+            The number of clusters, i.e. number of different types of 
+            genes
+
+        training_options : dict,default={}
+            The parameters for the inference scheme
+        """
+        self.K = int(K)
+        self.training_options = self._fill_training_options(training_options)
+
+    def fit(self, X, progress_bar=True):
+        """
+        Fits a model given count data X.
+
+        Parameters
+        ----------
+        X : np.array-like,shape=(N,p)
+            The count data
+        Returns
+        -------
+        self
+        """
+        td = self.training_options
+        K = self.K
+
+        N, p = X.shape
+
+        model = GaussianMixture(K)
+        model.fit(X)
+        Lambda_list_l = []
+        for i in range(K):
+            Lambda_list_l.append(
+                torch.tensor(
+                    1.2 * model.means_[i].astype(np.float32) + 0.1,
+                    requires_grad=True,
+                )
+            )
+
+        myrange = trange if progress_bar else range
+
+        pct_0 = np.mean(X == 0, axis=0)
+        pct_0 = 0.98 * pct_0 + 0.01
+        pct_0_inv = np.log(pct_0 / (1 - pct_0))
+
+        alpha_list_l = [
+            torch.tensor(pct_0_inv, requires_grad=True) for i in range(K)
+        ]
+
+        sigmoid = nn.Sigmoid()
+
+        data = torch.tensor(X.astype(np.float32))
+        pi_ = 1 / K * np.ones(K)
+        myloss = PoissonNLLLoss(log_input=False, full=True, reduction="none")
+
+        for i in myrange(td["n_iterations"]):
+            # E step compute most likely categories
+            Zprobs = np.zeros((N, K))
+            for k in range(K):
+                Lambda_k = Lambda_list_l[k]
+                alpha_k = 0.8 * sigmoid(alpha_list_l[k])  # Prob of being 0
```

Comment:
could you help me understand the need to use sigmoid here? just a knowledge gap on my part

Reply:
also where does 0.8 come from?

Reply:
Ah need to remove the 0.8, was a feasibility constraint to make sure no component has a zero probability inflation of more than 80%. As far as the reason for sigmoid, an easy way to do inference with gradient descent when you have variables with range not R is to make a latent variable with range R and map it to a feasible region using a transformation. In this case needs to be in the range [0,.8] which is what I have achieved.